### DISCUSSION TITLE
3 Band Equalizer not working

### DISCUSSION THREAD

**Discussion by greno** [pschatzmann/arduino-audiokit#63] [Q&A] [answered]
I'm getting the following error using the 3 band equalizer:

[E] Equilizer.h : 176 - Only 16 bits supported: 1073424384

It seems like the config is being overwritten in memory somewhere.   I still hear the input audio, but without the EQ applied.

Here is my test code:

```
// Audio
AudioKitStream kit; // Access I2S as stream
Equilizer3Bands eq(kit);
StreamCopy copier(kit, eq); // copy effects to kit

//Overall volume
float volumeControl = 0.98;

// Audio Format
const int sample_rate = 44100;
const int channels = 1;
const int bits_per_sample = 16;

void setup() {
  Serial.begin(115200);
  delay(800);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  auto cfg = kit.defaultConfig(RXTX_MODE);
  cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE2;

  // minimize lag
  cfg.buffer_count = 2;
  cfg.buffer_size = 256;

  kit.begin(cfg);
  
  kit.setVolume(volumeControl);
  es8388_set_mic_gain(MIC_GAIN_12DB);  

  // EQ
  auto eqcfg = eq.defaultConfig();
  eqcfg.gain_high = .9;
  eqcfg.gain_medium = .01;
  eqcfg.gain_low = .01;
  eqcfg.bits_per_sample = bits_per_sample;
  eqcfg.channels = channels;
  eqcfg.sample_rate = sample_rate;
  eq.begin(eqcfg);
}

void loop() {
  copier.copy();
  kit.processActions();
}
```

I added some extra debugging statements.  Here is the value from the setAudioInfo in Equilizer.h:  
`[W] Equilizer.h : 100 - GR - setAudioInfo - bits_per_sample: 16`

And here is the log messages from the filterSamples function, including my extra debug statement.  As you can see, the bits_per_sample value already has garbage in it on the first call. 
```

[W] AudioKit.h : 55 - sd_active = true
[W] AudioKit.h : 626 - Mode Button ignored because of conflict: 13 
[W] Equilizer.h : 161 - GR - filterSamples - bits_per_sample: 1073424384, sample rate: 3, channels: -2146597828
[E] Equilizer.h : 176 - Only 16 bits supported: 1073424384
[W] Equilizer.h : 161 - GR - filterSamples - bits_per_sample: 1073424384, sample rate: 1073492644, channels: -2146597828
[E] Equilizer.h : 176 - Only 16 bits supported: 1073424384
[W] Equilizer.h : 161 - GR - filterSamples - bits_per_sample: 1073424384, sample rate: 1073492644, channels: -2146597828
```

**Comment by pschatzmann** [comment, marked as answer]
This sketch has 2 mistakes: 
- make sure that you keep the number of channels consistent across all objects
- The sketch works with references so that you can dynamically change the parameters in the loop. So eigher store the config as global variable(like the provided example ) or if you work with temporary local variables define it as reference (  auto &eqcfg = eq.defaultConfig();).

**Reply by greno** [reply]
That makes sense, and making the cfgs global fixed that issue - thanks. 

Since the EQ is not an AudioEffect, how should EQ be used with an effect since both use a copier?  Is the right approach to use two copiers somehow? Or should EQ be converted to an AudioEffect?

**Comment by pschatzmann** [comment]
No need for an additional copy. 
You can specify the Equilizer3Bands instead of the I2SStream/AudioKitStream directly as ouput of the effects

**Reply by greno** [reply]
Perfect - thanks.
